File: patcha/utils/guard.py
import logging
from typing import Optional

logger = logging.getLogger(__name__)


class CollectorGuard:
    """Silences and disables a collector after max_failures consecutive failures.

    Usage:
        guard = CollectorGuard("safari", max_failures=3)

        if not guard.ok:
            return []
        try:
            ...
            guard.success()
        except Exception as e:
            guard.fail(e)
    """

    def __init__(self, name: str, max_failures: int = 3):
        self.name = name
        self.max_failures = max_failures
        self._failures = 0
        self._disabled = False

    @property
    def ok(self) -> bool:
        return not self._disabled

    def success(self) -> None:
        self._failures = 0

    def fail(self, exc: Exception, msg: Optional[str] = None) -> None:
        if self._disabled:
            return
        self._failures += 1
        label = msg or str(exc)
        if self._failures >= self.max_failures:
            self._disabled = True
            logger.warning(
                "%s: disabled after %d consecutive failures (last: %s). "
                "Will not retry until restart.",
                self.name,
                self._failures,
                label,
            )
        else:
            logger.warning(
                "%s: failure %d/%d — %s",
                self.name,
                self._failures,
                self.max_failures,
                label,
            )

File: rust/patcha/src/collectors/guard.rs
/// Disables a collector after `max_failures` consecutive errors.
pub struct CollectorGuard {
    pub name: String,
    max_failures: u32,
    failures: u32,
    disabled: bool,
}

impl CollectorGuard {
    pub fn new(name: impl Into<String>, max_failures: u32) -> Self {
        Self {
            name: name.into(),
            max_failures,
            failures: 0,
            disabled: false,
        }
    }

    pub fn ok(&self) -> bool {
        !self.disabled
    }

    pub fn success(&mut self) {
        self.failures = 0;
    }

    pub fn fail(&mut self, err: &anyhow::Error) {
        if self.disabled {
            return;
        }
        self.failures += 1;
        if self.failures >= self.max_failures {
            self.disabled = true;
            tracing::warn!(
                collector = %self.name,
                failures = self.failures,
                error = %err,
                "collector disabled after consecutive failures"
            );
        } else {
            tracing::warn!(
                collector = %self.name,
                failures = self.failures,
                error = %err,
                "collector error"
            );
        }
    }
}
